iothub/device/src/Transport/Amqp/AmqpUnit.cs

```diff
@@ -293,95 +328,109 @@ public void OnEventsReceived(Message message)
         #region Method
         public async Task EnableMethodsAsync(TimeSpan timeout)
         {
-            if (Logging.IsEnabled) Logging.Enter(this, timeout, $"{nameof(EnableMethodsAsync)}");
-
-            try
+            if (_closed)
             {
-                Debug.Assert(_methodSendingLink == null);
-                Debug.Assert(_methodReceivingLink == null);
-
-                string correlationIdSuffix = Guid.NewGuid().ToString();
-
-                Task<AmqpIoTReceivingLink> receiveLinkCreator = _amqpIoTSession.OpenMethodsReceiverLinkAsync(_deviceIdentity, correlationIdSuffix, timeout);
-                Task<AmqpIoTSendingLink> sendingLinkCreator = _amqpIoTSession.OpenMethodsSenderLinkAsync(_deviceIdentity, correlationIdSuffix, timeout);
-                await Task.WhenAll(receiveLinkCreator, sendingLinkCreator).ConfigureAwait(false);
-
-                _methodReceivingLink = receiveLinkCreator.Result;
-                _methodSendingLink = sendingLinkCreator.Result;
-
-                _methodReceivingLink.RegisterMethodListener(OnMethodReceived);
-                _methodSendingLink.Closed += OnLinkDisconnected;
-                _methodReceivingLink.Closed += OnLinkDisconnected;
+                throw DEVICE_DISCONNECTED_EXCEPTION;
+            }
 
-                if (Logging.IsEnabled) Logging.Associate(this, _methodReceivingLink, $"{nameof(_methodReceivingLink)}");
-                if (Logging.IsEnabled) Logging.Associate(this, _methodSendingLink, $"{nameof(_methodSendingLink)}");
+            if (Logging.IsEnabled) Logging.Enter(this, timeout, $"{nameof(EnableMethodsAsync)}");
+            AmqpIoTSession amqpIoTSession = await EnsureSessionAsync(timeout).ConfigureAwait(false);
+            bool gain = await _methodLinkLock.WaitAsync(timeout).ConfigureAwait(false);
+            if (!gain)
+            {
+                throw new TimeoutException();
             }
-            catch (Exception)
+            
+            string correlationIdSuffix = Guid.NewGuid().ToString();
+            try
             {
-                _methodReceivingLink?.Abort();
-                _methodReceivingLink = null;
-
-                _methodSendingLink?.Abort();
-                _methodSendingLink = null;
-
-                throw;
+                await Task.WhenAll(
+                    OpenMethodsReceiverLinkAsync(amqpIoTSession, correlationIdSuffix, timeout),
```

Comment:
How are these links cleaned if one of the open fails?

Reply:
By design - If a link opening fails, this try will throw exception and hence when this block of code is executed again, it will now try to open a new link for the failed link and use the already opened ;link fromt he previous execution.